metals/src/main/scala/scala/meta/internal/parsing/JavaFoldingRangeExtractor.scala

```diff
@@ -1,223 +1,230 @@
+    val iter = elems.iterator
+    // only one CompilationUnitTree for a single file
+    if (iter.hasNext) Some((trees, iter.next())) else None
+  }
 
-        case _ =>
-          acc.reverse
+  // search for any comments in file that aren't defined within a String
+  private def findComments(
+      text: String,
+      exclusions: List[Range],
+  ): List[Range] = {
+    @tailrec
+    def findComments(comments: List[Range], idx: Int): List[Range] = {
+      val startIdx = text.indexOf("/*", idx)
+      if (startIdx == -1)
+        comments
+      else {
+        val (newComments, newIdx) =
+          if (exclusions.exists(range => range.contains(startIdx.longValue)))
+            (comments, startIdx + 2)
+          else {
+            val endIdx = text.indexOf("*/", startIdx + 2)
+            val range = Range(
+              startIdx.longValue,
+              endIdx.longValue + 2,
+              FoldingRangeKind.Comment,
+            )
+            (range :: comments, endIdx + 2)
+          }
+        findComments(newComments, newIdx)
       }
-
     }
-    group(startAndEnds, Nil)
+
+    findComments(Nil, 0)
   }
 
-  private def createFoldingRange(
-      startLine: Int,
-      startCharacter: Int,
-      endLine: Int,
-      endCharacter: Int,
-      kind: String,
-  ): Option[FoldingRange] = {
-    if (endLine - startLine >= spanThreshold) {
-      val newFoldingRange = new FoldingRange(startLine, endLine)
-      newFoldingRange.setStartCharacter(startCharacter)
-      newFoldingRange.setEndCharacter(endCharacter)
-      newFoldingRange.setKind(kind)
-      Some(newFoldingRange)
-    } else {
-      None
+  def extract(
+      text: String,
+      path: AbsolutePath,
+      // TODO - what is this supposed to do???  Is it relevant in Java?
+      foldOnlyLines: Boolean,
+  ): List[FoldingRange] = {
+
+    getTrees(text, path) match {
+      case Some((trees, root)) =>
+        val sourcePositions = trees.getSourcePositions
+        val scanner = new FoldScanner(root, sourcePositions, text)
+        scanner.scan(root, {})
+        val lineMap = root.getLineMap
+        // imports are defined as a range per import but should be treated as one range encompassing all imports
+        val mergedImports = mergeRanges(
+          scanner.imports.map(range => range.toFoldingRange(lineMap))
+        )
+        // comments are not returned by the scanner so search separately
+        val comments = findComments(text, scanner.strings)
+        val allRanges = mergedImports :::
+          scanner.regions.map(range => range.toFoldingRange(lineMap)) :::
+          scanner.strings.map(range => range.toFoldingRange(lineMap)) :::
+          comments.map(range => range.toFoldingRange(lineMap))
+        allRanges
+          .filter(range =>
+            range.getEndLine - range.getStartLine >= spanThreshold
+          )
+      case None => Nil
     }
   }
 
+  private def mergeRanges(ranges: List[FoldingRange]): List[FoldingRange] = {
```

Comment:
Why not just find first and last range and use start and end from them?

Reply:
Done.  It's better as well as it will fold imports over comments